File: daily_problems/2026/06/0612/personal_submission/cf103102e_yawn_sean.cpp
#include <bits/stdc++.h>
// #pragma GCC optimize("O3,Ofast,unroll-loops")
// #include "atcoder/all"

using namespace std;
auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	int n;
	cin >> n;

	vector<vector<int>> cnt(3, vector<int>(3, 0));
	long long ans = 0;
	int cur1 = 0, cur2 = 0;

	cnt[0][0] = 1;

	while (n --) {
		int x;
		cin >> x;

		x %= 3;
		cur1 = (cur1 + x) % 3;
		cur2 = (cur2 + x * x) % 3;

		for (int i = 0; i < 3; i ++) {
			for (int j = 0; j < 3; j ++) {
				int v1 = (cur1 - i) % 3;
				int v2 = (cur2 - j) % 3;

				if ((v1 * v1 - v2) % 3 == 0) ans += cnt[i][j];
			}
		}

		cnt[cur1][cur2] ++;
	}

	cout << ans;

	return 0;
}

File: daily_problems/2026/06/0612/personal_submission/cf103102e_yawn_sean.py
# Submission link: https://codeforces.com/gym/103102/submission/378142095
def main():
    n = II()
    nums = LII()
    
    ans = 0
    
    cnt = [[0] * 3 for _ in range(3)]
    cnt[0][0] = 1
    
    cur1 = 0
    cur2 = 0
    
    for x in nums:
        x %= 3
        cur1 = (cur1 + x) % 3
        cur2 = (cur2 + x * x) % 3
        
        for i in range(3):
            for j in range(3):
                v1 = (cur1 - i) % 3
                v2 = (cur2 - j) % 3
                
                if v1 * v1 % 3 == v2:
                    ans += cnt[i][j]
        
        cnt[cur1][cur2] += 1
    
    print(ans)